### PR TITLE
Stabilize and simplify SIL linkage and serialization

### PR DESCRIPTION
**Pull request by eeckstein**
The main point of this change is to make sure that a shared function always has a body: both, in the optimizer pipeline and in the swiftmodule file.
This is important because the compiler always needs to emit code for a shared function. Shared functions cannot be referenced from outside the module.
In several corner cases we missed to maintain this invariant which resulted in unresolved-symbol linker errors.

As side-effect of this change we can drop the shared_external SIL linkage and the IsSerializable flag, which simplifies the serialization and linkage concept.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci Please test compiler performance

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci Please test compiler performance

**Comment by eeckstein** [comment]
@swift-ci test macOS

**Comment by eeckstein** [comment]
@swift-ci test macOS

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci test

File: include/swift/SIL/SILLinkage.h [hunk 1]
```diff
@@ -38,33 +38,51 @@ enum class SILLinkage : uint8_t {
   /// This object definition is visible to multiple Swift modules (and
   /// thus potentially across linkage-unit boundaries).  There are no
   /// other object definitions with this name in the program.
+  ///
+  /// Public functions must be definitions, i.e. must have a body, except they
+  /// are clang imported.
```
Comment (slavapestov) [inline include/swift/SIL/SILLinkage.h:43]: Nitpick -- s/clang imported/body emitted by clang/? This is only true in a few narrow cases like bitfield accessors and static inline functions IIRC

**Review by slavapestov — APPROVED** [review]
This is a nice conceptual simplification, thank you!

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci test

**Pull request merged**